File: JAVA/SEARCHING_ALGORITHM/BFS.java
import java.util.*;

class BFS {
    private int numberOfVertices; // Number of vertices in the graph
    private LinkedList<Integer>[] adjacencyList; // Array of LinkedLists to store adjacency list representation of the graph

    // Constructor to initialize the graph
    Graph(int vertices) {
        numberOfVertices = vertices; // Set the number of vertices
        adjacencyList = new LinkedList[vertices]; // Create an array of linked lists for each vertex
        for (int i = 0; i < vertices; i++) {
            adjacencyList[i] = new LinkedList<>(); // Initialize each linked list
        }
    }

    // Method to add an edge to the graph (from source to destination)
    void addEdge(int source, int destination) {
        adjacencyList[source].add(destination); // Add destination to the adjacency list of the source vertex
        // For undirected graph, we would also add source to destination's list (adjacencyList[destination].add(source))
    }

    // Method to perform BFS traversal starting from the given startVertex
    void bfs(int startVertex) {
        // Array to track whether a vertex has been visited (initialized to false by default)
        boolean[] visited = new boolean[numberOfVertices];

        // Queue to help with the BFS traversal (FIFO structure)
        Queue<Integer> queue = new LinkedList<>();

        // Start BFS from the given startVertex
        visited[startVertex] = true; // Mark the start vertex as visited
        queue.add(startVertex); // Enqueue the start vertex

        // Continue BFS until the queue is empty
        while (!queue.isEmpty()) {
            // Dequeue the current vertex from the front of the queue
            int currentVertex = queue.poll();
            System.out.print(currentVertex + " "); // Print the current vertex

            // Get all adjacent vertices (neighbors) of the current vertex
            for (int neighbor : adjacencyList[currentVertex]) {
                // If the adjacent vertex hasn't been visited yet
                if (!visited[neighbor]) {
                    visited[neighbor] = true; // Mark it as visited
                    queue.add(neighbor); // Enqueue the vertex to explore it in subsequent iterations
                }
            }
        }
    }

    // Main method to test the BFS implementation
    public static void main(String[] args) {
        // Create a graph with 6 vertices (0 to 5)
        Graph graph = new Graph(6);

        // Add edges to the graph (directed edges)
        graph.addEdge(0, 1); // Edge from vertex 0 to vertex 1
        graph.addEdge(0, 2); // Edge from vertex 0 to vertex 2
        graph.addEdge(1, 3); // Edge from vertex 1 to vertex 3
        graph.addEdge(1, 4); // Edge from vertex 1 to vertex 4
        graph.addEdge(2, 4); // Edge from vertex 2 to vertex 4
        graph.addEdge(3, 5); // Edge from vertex 3 to vertex 5
        graph.addEdge(4, 5); // Edge from vertex 4 to vertex 5

        // Perform BFS starting from vertex 0
        System.out.println("Breadth-First Search starting from vertex 0:");
        graph.bfs(0);
    }
}

File: C++/graph/BFS.cpp
// Here’s how you can implement Breadth-First Search (BFS) in C++ using the graph structure

#include <iostream>
#include <unordered_map>
#include <list>
#include <queue>
#include <vector>

class Graph {
private:
    // Adjacency list to store the graph
    std::unordered_map<int, std::list<int>> adjList;

public:
    // Function to add an edge between two vertices
    void addEdge(int v1, int v2) {
        adjList[v1].push_back(v2);
        adjList[v2].push_back(v1);  // Since the graph is undirected
    }

    // Function to perform BFS traversal from a given source vertex
    void BFS(int start) {
        // Vector to track visited vertices
        std::vector<bool> visited(adjList.size(), false);
        // Queue to help with the BFS traversal
        std::queue<int> q;

        // Start from the given vertex
        visited[start] = true;
        q.push(start);

        // Continue the BFS until the queue is empty
        while (!q.empty()) {
            // Get the next vertex from the queue
            int v = q.front();
            q.pop();

            std::cout << v << " ";

            // Get all the adjacent vertices of the dequeued vertex
            for (int neighbor : adjList[v]) {
                if (!visited[neighbor]) {
                    // Mark neighbor as visited and enqueue it
                    visited[neighbor] = true;
                    q.push(neighbor);
                }
            }
        }
    }
};

int main() {
    Graph g;

    // Adding edges
    g.addEdge(0, 1);
    g.addEdge(0, 2);
    g.addEdge(1, 2);
    g.addEdge(1, 3);
    g.addEdge(2, 3);
    g.addEdge(3, 4);

    // Perform BFS starting from vertex 0
    std::cout << "BFS starting from vertex 0:" << std::endl;
    g.BFS(0);

    return 0;
}